lms/product/plugin/plugin.py

```diff
@@ -0,0 +1,34 @@
+from dataclasses import dataclass
+
+from lms.product.plugin.grouping_service import GroupingServicePlugin
+
+
+@dataclass
+class PluginConfig:
+    """A collection of plugin class definitions."""
+
+    # These also provide the default implementations
+    grouping_service: type = GroupingServicePlugin
+
+
+class Plugins:
+    """A collection of plugins used to separate LMS specific functionality."""
+
+    class _LazyPlugin:
+        # Lazy load plugins based on the plugin config from pyramid services
+        plugin_name = None
+
+        def __set_name__(self, owner, name):
+            self.plugin_name = name
+
+        def __get__(self, instance, owner):
+            plugin_class = getattr(instance._plugin_config, self.plugin_name)
+            plugin = instance._request.find_service(iface=plugin_class)
+            setattr(instance, self.plugin_name, plugin)  # Overwrite the attr
+            return plugin
```

Comment:
This is the only bit of complexity here really. This is a descriptor. When Python adds this to the class `__set_name__` is called and we are told the name of the attr we were called with. We use this to find out the correct plugin class from the config and look it up and return it.

We also overwrite the attr which means subsequent access is as fast as it can be.